### DISCUSSION TITLE
Keep track of cancelled timer?

### DISCUSSION THREAD

**Discussion by ferbulous** [arendst/Tasmota#21878] [Q&A] [answered]
Hi, could I keep track of cancelled timer?
I’d like to run an action by cancelling the timer repeatedly.
Something like this, basically turning the relay off and on (in less than 2 seconds) triggers the timer which changes the color temperature.

``` 
On boot do var2 0 endon
On power2#state=0 do ruletimer1 2 endon
On power2#state=1 do ruletimer1 0 endon
On rules#timer!=1 do add1 2 endon //this isn’t correct but basically if the timer gets cancelled
On var2#state==1 do cmnd/bulb/ct 500 endon
On var2#state==2 do cmnd/bulb/ct 300 endon
On var2#state==3 do cmnd/bulb/ct 155 endon
On var2#state==4 do var2 1 endon

```

**Comment by sfromis** [comment]
Better explain your use case, instead asking for something which is not a feature.

**Reply by barbudor** [reply]
As you happen to cancel the ruletimer only in the 3rd rule, why not doing the 'add1 2' in the same using a backlog ?

**Reply by ferbulous** [reply]
My wife’s gotten used to using lights which can change to 3 preset color temperature by toggling the switch on and off.
I plan to replace them with smart lights (also tasmota) and I want to mimic the same toggling on/off behavior for her.
I’m aware there is multi press but these 
are just toggle switches with relays installed.
The relay to the light will be set to output high with fake relay similar to this

https://youtu.be/52ZPROouhOQ?si=jtC4621BRM5vxskc

**Reply by ferbulous** [reply]
> As you happen to cancel the ruletimer only in the 3rd rule, why not doing the 'add1 2' in the same using a backlog ?

Like this?
``` 
On boot do var2 0 endon
On power2#state=0 do add2 1 endon
On var2#state==1 do cmnd/bulb/ct 500 endon
On var2#state==2 do cmnd/bulb/ct 300 endon
On var2#state==3 do cmnd/bulb/ct 155 endon
On var2#state==4 do var2 1 endon
``` 

So i think i can just have it do that when it’s toggled off. Is it possible to add another condition to add to var2 only if it’s toggled off in less than 2 seconds for example?

**Reply by barbudor** [reply]
As you actually not using the timer at all (that I failed to noticed initialy), yes

Also,
- your syntax for the 'do' is wrong. Just use the CT command. You are mixing up commands a'd MQTT topics
- your syntax for the 'on boot' is wrong. Look at the doc to choose a boot trigger suit able for you. 'system#boot' should be ok

For your other question, you can probably use '%var3%' as the incrément instead of hard-coded '1' value. Along with a ruletimer.
But as I can't see in your rules how you turn off the light, I can't advice where to do that

**Reply by ferbulous** [reply]
Sorry about that, missed some lines

Using this as my reference

https://youtu.be/WAdsWeWvQL4?si=1fwrOSeSuobIrRfw

``` 
On system#boot do var2 0 endon
On power2#state=0 do add2 1 endon
On var2#state==1 do publish cmnd/bulb/ct 500 endon
On var2#state==2 do publish cmnd/bulb/ct 300 endon
On var2#state==3 do publish cmnd/bulb/ct 155 endon
On var2#state==4 do var2 1 endon
``` 
I’m not quite sure how to add the ruletimer so turning off less than 2 seconds would add var2.

**Comment by ferbulous** [comment]
Hi, I'm trying to revisit this again but with a different approach of simply checking if the timer is still running

```
on power1#state=0 do ruletimer1 3 endon
on power1#state=1 do IF ((%timer1%!=0) AND (%CT%<=250)) CT 500 else CT 160 endif endon 
```

Basically when the relay is turned off starts 3 second timer, and when turned on checks if the timer1 isn't finished and compares the color temperature.
I don't think I'm getting this right ```(%timer1%!=0)```, is there a proper way to check if the timer is still running?

**Reply by barbudor** [reply, marked as answer]
What you missed is that there are 2 type of 'timers' in Tasmota: RuleTimers as you are using above and the more generic Timers that are used to trigger actions at time-of-day.
The variable `%timerx%` refer to the later

The only way to get the value for a RuleTimer is from a rule to issue the command `Backlog RuleTimer` and to catch the value with `on T1 do`

**Reply by ferbulous** [reply]
Hi, do you mind sharing an example to issue the command with backlog and catching the value?
I couldn’t find it on the rules page

**Reply by ferbulous** [reply]
I added another line but it’s not updating var2, and I’m not exactly sure where to use backlog ruletimer here

```
on power1#state=0 do ruletimer1 3 endon
on T1#state do var2 %value% endon
on power1#state=1 do IF ((%var2%!=0) AND (%CT%<=250)) CT 500 else CT 160 endif endon 
```

**Reply by sfromis** [reply]
Did you try doing what barbudor suggested? Using `on T1 do` instead of `on T1#state do`? And `Backlog RuleTimer`, instead of just `RuleTimer`.

**Reply by ferbulous** [reply]
I did try ``` on T1 do ``` earlier but no difference
So it should be like this?

``` 
on power1#state=0 do backlog ruletimer1 3 endon
on T1 do var2 %value% endon
on power1#state=1 do IF ((%var2%!=0) AND (%CT%<=250)) CT 500 else CT 160 endif endon 
```

**Reply by sfromis** [reply]
Yeah, you need to be doing both at the same time. Note that what you get in `var2` is only at the time of you issuing the `RuleTimer` command, it will only be updated if/when you issue the command again.

**Reply by ferbulous** [reply]
If added the timer1 (3 seconds) value for var2
Wouldn’t it still update var2 each second it’s counting down?

**Reply by sfromis** [reply]
No, I really meant what I wrote:
> it will only be updated if/when you issue the command again.

When triggers are based on the JSON you get in the console (and MQTT), they can only fire when getting the JSON payload.

**Reply by ferbulous** [reply]
Thanks guys, finally got it working
Always thought you only need backlog to group multiple command without restart

```
Rule1
on power1#state=0 do backlog ruletimer1 5 endon 
on T1 do var2 %value% endon 
on Rules#Timer=1 do var2 0 endon
on CT do var3 %value% endon
on power1#state=1 do IF ((%var2%!=0) AND (%var3%<=250)) CT 490 ELSE CT 170 ENDIF endon
```

Added ```Rules#Timer=1``` so var2 gets 0 if lights are turned off  longer than 5 seconds. That way it doesn't run the color temperature change command next time it's turned on

**Reply by barbudor** [reply]
I see nowhere you are issueing a `Backlog RulesTimer` to get the current value of RuleTimer1 at the moment you need it, that is when Power is switched to 1
With your current rules, var2 only get the value of RulesTimer1 when you start it: value is 5 until the timer expires and set var2 to 0

 A possible solution could be (not tested):
```
Rule1
on power1#state=0 do ruletimer1 5 endon 
on T1 do var2 %value% endon 
on Rules#Timer=1 do var2 0 endon
on CT do var3 %value% endon
on power1#state=1 do backlog ruletimer; event test endon
on event#test do IF ((%var2%!=0) AND (%var3%<=250)) CT 490 ELSE CT 170 ENDIF endon
```

But there's always a risk of race condition near 0
